server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4772,12 +4784,40 @@ private String copySegmentFiles(
             for (String file : uploadedSegments.keySet()) {
                 long checksum = Long.parseLong(uploadedSegments.get(file).getChecksum());
                 if (overrideLocal || localDirectoryContains(storeDirectory, file, checksum) == false) {
+                    segmentsToDownload.add(file);
+                    totalSizeOfSegmentsToDownload += uploadedSegments.get(file).getLength();
+                } else {
+                    skippedSegments.add(file);
+                }
+            }
+            // Upload stats before starting Segment downloads
+            // Adding a check to prevent stats being published for NoOp runs
+            if (!segmentsToDownload.isEmpty()) {
+                beforeSegmentDownloads(downloadStatsTracker, totalSizeOfSegmentsToDownload);
+            }
+            long sizeOfDownloadedSegments = 0;
+            try {
+                long startTimeInMs = System.currentTimeMillis();
+                // Copying segments files to local store directory from remote store directory (Shard Recovery)
+                for (String file : segmentsToDownload) {
+                    // Throws IOException if download fails
                     storeDirectory.copyFrom(sourceRemoteDirectory, file, file, IOContext.DEFAULT);
                     storeDirectory.sync(Collections.singleton(file));
                     downloadedSegments.add(file);
-                } else {
-                    skippedSegments.add(file);
+                    sizeOfDownloadedSegments += uploadedSegments.get(file).getLength();
```

Comment:
We should be starting track the download bytes started/succeeded/failed within this method.